File: Rust/fft.rs
use std;

// Internal variables
static mut PHASEVEC_EXIST : bool = false;
static mut PHASEVEC: [[f64; 2]; 32] = [[0.0; 2]; 32];

// Public function
pub unsafe fn fft(log2point: u32, xy_in: [[f64; 2]; 4096]) -> [[f64; 2]; 4096] {
    let mut xy_out: [[f64; 2]; 4096] = [[0.0; 2]; 4096];

    if !PHASEVEC_EXIST {;
	for i in 0..32 {
	    let point: i32 = 2<<i;
	    PHASEVEC[i][0] = (-2.0*std::f64::consts::PI/(point as f64)).cos();
	    PHASEVEC[i][1] = (-2.0*std::f64::consts::PI/(point as f64)).sin();
	}
	PHASEVEC_EXIST = true;
    };
    for i in 0 .. 1<<log2point {
	let mut brev: usize = i;
	brev = ((brev & 0xaaaaaaaa) >> 1) | ((brev & 0x55555555) << 1);
	brev = ((brev & 0xcccccccc) >> 2) | ((brev & 0x33333333) << 2);
	brev = ((brev & 0xf0f0f0f0) >> 4) | ((brev & 0x0f0f0f0f) << 4);
	brev = ((brev & 0xff00ff00) >> 8) | ((brev & 0x00ff00ff) << 8);
	brev = (brev >> 16) | (brev << 16);

	brev >>= 32-log2point;
	xy_out[brev] = xy_in[i];
    }

    // here begins the Danielson-Lanczos section;
    let n: usize = 1<<log2point;
    let mut l2pt: usize = 0;
    let mut mmax: usize = 1;

    while n > mmax {;
	let istep: usize = mmax<<1;
//	double theta = -2*M_PI/istep;
//	double complex wphase_xy = cos(theta) + sin(theta)*I;
	let wphase_xy: [f64; 2] = PHASEVEC[l2pt];
	l2pt+=1;
	let mut w_xy: [f64; 2] = [1.0, 0.0];
	for m in 0..mmax {;
//	    for i in (m .. n).step_by(istep) { -- now unstable
	    let mut i = m;
	    while i < n {
//		let temp_xY: [f64; 2] = w_xy * xy_out[i+mmax]; -- missing complex arithmetic
//		xy_out[i+mmax]  = xy_out[i] - temp_xY;
//		xy_out[i     ] += temp_xY;
		let temp_x: f64 = w_xy[0] * xy_out[i+mmax][0] - w_xy[1] * xy_out[i+mmax][1];
		let temp_y: f64 = w_xy[0] * xy_out[i+mmax][1] + w_xy[1] * xy_out[i+mmax][0];
		xy_out[i+mmax][0]  = xy_out[i][0] - temp_x;
		xy_out[i+mmax][1]  = xy_out[i][1] - temp_y;
		xy_out[i     ][0] += temp_x;
		xy_out[i     ][1] += temp_y;
	        i += istep;
	    }
//	    w_xy *= wphase_xy; // rotate -- missing complex arithmetic
	    let w_tmp = w_xy[0] * wphase_xy[0] - w_xy[1] * wphase_xy[1];
	    w_xy[1] = w_xy[0] * wphase_xy[1] + w_xy[1] * wphase_xy[0];
	    w_xy[0] = w_tmp;
	}
	mmax=istep;
    }
    return xy_out;
}

File: Java/Fft.java
public class Fft {

    // Internal variables
    static int phasevec_exist = 0;
    static double[][] phasevec = new double[32][2];

    // Public function
    public static void fft(int log2point, double [][] xy_out, double [][] xy_in) {
	if (phasevec_exist == 0) {
	    for (int i=0; i<32; i++) {
		int point = 2<<i;
		phasevec[i][0] = Math.cos(-2*Math.PI/point);
		phasevec[i][1] = Math.sin(-2*Math.PI/point);
	    }
	    phasevec_exist = 1;
	}
	for (int i=0; i < (1<<log2point); i++) {
	    long brev = i;
	    brev = ((brev & 0xaaaaaaaa) >> 1) | ((brev & 0x55555555) << 1);
	    brev = ((brev & 0xcccccccc) >> 2) | ((brev & 0x33333333) << 2);
	    brev = ((brev & 0xf0f0f0f0) >> 4) | ((brev & 0x0f0f0f0f) << 4);
	    brev = ((brev & 0xff00ff00) >> 8) | ((brev & 0x00ff00ff) << 8);
	    brev = (brev >> 16) | (brev << 16);

	    brev >>= 32-log2point;
	    xy_out[(int)brev] = xy_in[i].clone();
	}

	// here begins the Danielson-Lanczos section
	int n = 1<<log2point;
	int l2pt=0;
	int mmax=1;

	while (n>mmax) {
	    int istep = mmax<<1;

//	    double theta = -2*M_PI/istep;
//	    double complex wphase_XY = cos(theta) + sin(theta)*I;
	    double[] wphase_XY = phasevec[l2pt++];
	    double[] w_XY = {1.0, 0.0};

	    for (int m=0; m < mmax; m++) {
		for (int i=m; i < n; i += istep) {
		    double[] tempXY = { w_XY[0] *xy_out[i+mmax][0] - w_XY[1] *xy_out[i+mmax][1],
					w_XY[0] *xy_out[i+mmax][1] + w_XY[1] *xy_out[i+mmax][0] };

		    xy_out[i+mmax][0]  = xy_out[i][0] - tempXY[0];
		    xy_out[i+mmax][1]  = xy_out[i][1] - tempXY[1];

		    xy_out[i     ][0] += tempXY[0];
		    xy_out[i     ][1] += tempXY[1];
		}
		double w_X_tmp;
		w_X_tmp = w_XY[0] * wphase_XY[0] - w_XY[1] * wphase_XY[1]; // rotate
		w_XY[1] = w_XY[0] * wphase_XY[1] + w_XY[1] * wphase_XY[0]; // rotate
		w_XY[0] = w_X_tmp;
	    }
	    mmax=istep;
	}
    }
}
